### PR TITLE
[SPARK-16959] [SQL] Rebuild Table Comment when Retrieving Metadata from Hive Metastore

### PR DESCRIPTION
**Pull request by gatorsmile**
### What changes were proposed in this pull request?

The `comment` in `CatalogTable` returned from Hive is always empty. We store it in the table property when creating a table. However, when we try to retrieve the table metadata from Hive metastore, we do not rebuild it. The `comment` is always empty.

This PR is to fix the issue.
### How was this patch tested?

Fixed the test case to verify the change.

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @cloud-fan . This is a tiny fix. Could you review this?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 1]
```diff
@@ -135,7 +135,9 @@ class HiveDDLSuite
         sql(s"CREATE VIEW $viewName COMMENT 'no comment' AS SELECT * FROM $tabName")
         val tableMetadata = catalog.getTableMetadata(TableIdentifier(tabName, Some("default")))
         val viewMetadata = catalog.getTableMetadata(TableIdentifier(viewName, Some("default")))
+        assert(tableMetadata.comment == Option("BLABLA"))
         assert(tableMetadata.properties.get("comment") == Option("BLABLA"))
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: I think we should also remove the `comment` from table properties, to not surprise users.

**Comment by cloud-fan** [comment]
do you know why we put comment in table properties instead of table comment field? Some hive tricks?

**Comment by gatorsmile** [comment]
After an investigation, Hive does not have such a field. Thus, they also put it in the table properties/parameters. 

For example, this is the output of one table with table comment `d`
```
hive> desc formatted tab5;
OK
# col_name            	data_type           	comment             
	 	 
id                  	int                 	b                   
	 	 
# Detailed Table Information	 	 
Database:           	default             	 
Owner:              	root                	 
CreateTime:         	Mon Aug 08 16:19:12 EDT 2016	 
LastAccessTime:     	UNKNOWN             	 
Protect Mode:       	None                	 
Retention:          	0                   	 
Location:           	hdfs://6b68a24121f4:9000/user/hive/warehouse/tab5	 
Table Type:         	MANAGED_TABLE       	 
Table Parameters:	 	 
	comment             	d                   
	transient_lastDdlTime	1470687552          
	 	 
# Storage Information	 	 
SerDe Library:      	org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe	 
InputFormat:        	org.apache.hadoop.mapred.TextInputFormat	 
OutputFormat:       	org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat	 
Compressed:         	No                  	 
Num Buckets:        	-1                  	 
Bucket Columns:     	[]                  	 
Sort Columns:       	[]                  	 
Storage Desc Params:	 	 
	serialization.format	1     
```

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala (hunk 1 above)
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: As explained below, Hive keeps `comment` in the table properties. Should we keep it too?

Thanks!

Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: I don't think so. `HiveClient` should be symmetrical, the table meta read back should be same with what users saved into. We don't need to follow hive here.

Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: Ok. I see. Let me fix it now.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 2]
```diff
@@ -413,7 +413,8 @@ private[hive] class HiveClientImpl(
           properties = Option(h.getTTable.getSd.getSerdeInfo.getParameters)
             .map(_.asScala.toMap).orNull
         ),
-        properties = properties,
+        properties = properties.filter(kv => kv._1 != "path"),
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: why not use `filterKeys`?

Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: The implementation of `filterKeys` will create a new object.
```
override def filterKeys(p: A => Boolean): Map[A, B] = new FilteredKeys(p) with DefaultMap[A, B]
```
Thus, I got the following error:
``` org.apache.spark.SparkException: Task not serializable```

Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: ah i see

Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: hey, we should filter out `comment` here, not `path`

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 3]
```diff
@@ -135,8 +135,8 @@ class HiveDDLSuite
         sql(s"CREATE VIEW $viewName COMMENT 'no comment' AS SELECT * FROM $tabName")
         val tableMetadata = catalog.getTableMetadata(TableIdentifier(tabName, Some("default")))
         val viewMetadata = catalog.getTableMetadata(TableIdentifier(viewName, Some("default")))
-        assert(tableMetadata.properties.get("comment") == Option("BLABLA"))
-        assert(viewMetadata.properties.get("comment") == Option("no comment"))
+        assert(tableMetadata.comment == Option("BLABLA"))
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: also test that the `comment` is not in table properties?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: haha, sure. Multiple PRs are being done at the same time. Will open a new PR tomorrow regarding `path`...

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala (hunk 3 above)
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: Added the test. : )

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by gatorsmile** [comment]
@cloud-fan To backport #14531, we need to backport this. Do you want me to do it?

**Comment by cloud-fan** [comment]
yea, go ahead :)

**Comment by cloud-fan** [comment]
backported to 2.0

**Comment by gatorsmile** [comment]
I need to fix the test case failure in Branch 2.0

**Pull request closed without merging**